src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala

```diff
@@ -47,82 +45,12 @@ private[tracker] object TaskOpProcessorImpl {
-          Action.Noop
-        }
-    }
-    private[this] def resolveForResidentFailed(task: Task, status: TaskStatus): Option[Action] = {
-      for {
-        _ <- task.reservationWithVolumes
-        _ <- task.launched
-        if Terminated.isTerminated(status.getState)
-      } yield Action.Update(task.copy(launched = None))
-    }
-
-    /**
-      * Calculates the change that needs to performed on this task according to the given task status update
-      */
-    private[this] def resolveForLaunchedTask(task: Task, statusUpdate: TaskStatus): Option[Action] = {
-      task.launched.map { launched =>
-        statusUpdate.getState match {
-          case Terminated(_) => Action.Expunge
-          case TASK_RUNNING if !launched.hasStartedRunning => stagedNowRunning(task, launched, statusUpdate)
-          case _ => updateTaskOnStateChange(task, launched, statusUpdate)
-        }
-      }
-    }
-
-    private[this] def stagedNowRunning(task: Task, currentLaunched: Task.Launched, statusUpdate: TaskStatus): Action = {
-      val now = clock.now()
-      Action.Update(
-        task.copy(launched = Some(
-          currentLaunched.copy(
-            status = currentLaunched.status.copy(
-              startedAt = Some(now),
-              mesosStatus = Some(statusUpdate)
-            )
-          )
-        ))
-      )
-    }
-    private[this] def updateTaskOnStateChange(
-      taskState: Task, currentLaunch: Task.Launched, statusUpdate: TaskStatus): Action = {
-
-      def updatedOnChange(currentStatus: TaskStatus): Option[Task] = {
-        val healthy =
-          statusUpdate.hasHealthy && (!currentStatus.hasHealthy || currentStatus.getHealthy != statusUpdate.getHealthy)
-        val changed = healthy || currentStatus.getState != statusUpdate.getState
-        if (changed) {
-          Some(
-            taskState.copy(
-              launched = Some(
-                currentLaunch.copy(
-                  status = currentLaunch.status.copy(mesosStatus = Some(currentStatus))
-                )
-              )
-            )
-          )
-        }
-        else {
-          log.info("currentStatus {}", currentStatus)
-          log.info("update {}", statusUpdate)
-          None
-        }
-      }
-
-      val maybeUpdated = currentLaunch.status.mesosStatus.flatMap(updatedOnChange(_))
-
-      maybeUpdated match {
-        case Some(updated) => Action.Update(updated)
-        case None =>
-          log.debug(s"Ignoring status update for ${taskState.taskId}. Status did not change.")
-          Action.Noop
+      val change = task.update(TaskStateOp.MesosUpdate(MarathonTaskStatus(statusUpdate), clock.now()))
+      change match {
+        case TaskStateChange.Update(updatedTask) => Action.Update(updatedTask)
+        case TaskStateChange.Expunge             => Action.Expunge
+        case TaskStateChange.NoChange            => Action.Noop
+        case TaskStateChange.Failure(_)          => Action.Noop
```

Comment:
Why not map to Action.Failure? And why mapping this at all and not just use Action. ...